src/engine.rs

```diff
@@ -365,49 +408,101 @@ impl Engine {
-                    let bind_group = self.bind_map.create_bind_group(
-                        device,
-                        &shader.bind_group_layout,
-                        bindings,
-                        external_resources,
-                        &mut self.pool,
-                    )?;
-                    let mut cpass = encoder.begin_compute_pass(&Default::default());
-                    #[cfg(feature = "wgpu-profiler")]
-                    profiler.begin_scope(shader.label, &mut cpass, device);
-                    cpass.set_pipeline(&shader.pipeline);
-                    cpass.set_bind_group(0, &bind_group, &[]);
-                    cpass.dispatch_workgroups(wg_size.0, wg_size.1, wg_size.2);
-                    #[cfg(feature = "wgpu-profiler")]
-                    profiler.end_scope(&mut cpass);
+                    if let Some(cpu_shader) = shader.cpu_shader {
+                        let resources =
+                            transient_map.create_cpu_resources(&mut self.bind_map, bindings);
+                        cpu_shader(wg_size.0, &resources);
+                    } else {
+                        let bind_group = transient_map.create_bind_group(
+                            &mut self.bind_map,
+                            &mut self.pool,
+                            device,
+                            queue,
+                            &mut encoder,
+                            &shader.bind_group_layout,
+                            bindings,
+                        )?;
+                        let mut cpass = encoder.begin_compute_pass(&Default::default());
+                        #[cfg(feature = "wgpu-profiler")]
+                        profiler.begin_scope(shader.label, &mut cpass, device);
+                        cpass.set_pipeline(&shader.pipeline);
+                        cpass.set_bind_group(0, &bind_group, &[]);
+                        cpass.dispatch_workgroups(wg_size.0, wg_size.1, wg_size.2);
+                        #[cfg(feature = "wgpu-profiler")]
+                        profiler.end_scope(&mut cpass);
+                    }
+                }
+                Command::DispatchIndirect(shader_id, proxy, offset, bindings) => {
+                    let shader = &self.shaders[shader_id.0];
+                    if let Some(cpu_shader) = shader.cpu_shader {
+                        let n_wg;
+                        if let CpuBinding::BufferRW(b) = self.bind_map.get_cpu_buf(proxy.id) {
+                            let slice = b.borrow();
+                            let indirect: &[u32] = bytemuck::cast_slice(&slice);
+                            n_wg = indirect[0];
+                        } else {
+                            panic!("indirect buffer missing from bind map");
+                        }
+                        let resources =
+                            transient_map.create_cpu_resources(&mut self.bind_map, bindings);
+                        cpu_shader(n_wg, &resources);
```

Comment:
Similar comment as above. In particular, I would mention that the indirect buffer is expected to have been initialized by a preceding CPU stage.

A question that comes to mind is why you would need indirect dispatch at all when all prior stages execute on the CPU, which means the workgroup count is known at this stage. I think it's worth clarifying in a comment that this system exists primarily for testing.

Reply:
The reasoning is to make CPU stages have exactly the same interfaces as the GPU counterparts. Obviously if you just want to run the CPU shaders to get results, indirect dispatch is a [karotousen](https://www.tokyoweekender.com/art_and_culture/japanese-culture/weekly-japanese-idiom-karotousen-useless-things/).

I'm in favor of clearly expressing goals, but am less sure that a comment in the code is the best way to do it.

Reply:
The comments you added look great. The broader design could probably use more documentation in a README.md or module-level comments but that can all wait until this is more hashed out.